### PR TITLE
Add background images

### PR DESCRIPTION
**Pull request by liwii**
## Context
In the current `zen-desktop`, the background is filled with one green-ish color. With a background image, the system would be more attractive to the users.
## Summary
Install the background images provided by @rtr-dnd and draw it on the screen. If the screen size ratio does not match the image size ratio, the image is scaled with aspect-fill manner.
image_123927839
## How to check behavior
Run `zen-desktop -s weston-terminal` and see if the background image is properly set. You can also test the other bundled background images by changing the line in https://github.com/zigen-project/zen/compare/main...liwii:bg?expand=1#diff-d6c974d638e0e2ea1f88ab4bb8e67b78b22a6aa636536a72467d9b39bb9defe0R14.  
If needed, change the screen orientation and the scale like the comment in #121 and test it again.

### CODE REVIEW

File: zen/screen-renderer.c [hunk 1]
```diff
@@ -24,6 +24,46 @@ struct surface_iterator_data {
...
+  float matrix[9];
+  int output_width, output_height;
+  struct wlr_output *wlr_output = screen->output->wlr_output;
+  wlr_output_transformed_resolution(wlr_output, &output_width, &output_height);
+  zn_screen_renderer_get_background_box(
+      &box, bg_texture, output_width, output_height);
+  wlr_matrix_project_box(matrix, &box, WL_OUTPUT_TRANSFORM_NORMAL, 0,
+      wlr_output->transform_matrix);
+  wlr_render_texture_with_matrix(renderer, bg_texture, matrix, 1.0f);
+}
```
Comment (watasuke102) [inline zen/screen-renderer.c:65]: When loading background image file is failed in some reasons (ex. file not found), `bg_texture` becomes NULL, and crash zen on rendering. Please check `bg_texture` at first.

**Review by watasuke102 — CHANGES_REQUESTED** [review]

File: zen/screen-renderer.c (hunk 1 above)
Comment (liwii) [reply zen/screen-renderer.c:66]: Fixed, thanks for the catch!!

**Review by watasuke102 — DISMISSED** [review]
Nice change!

File: zen/screen-renderer.c [hunk 2]
```diff
@@ -24,6 +24,47 @@ struct surface_iterator_data {
...
+    box->height = output_height;
+    box->width =
+        (int)(bg_texture->width * output_height / (double)bg_texture->height);
+    box->x = (output_width - box->width) / 2;
+  } else {
+    // Fit the width of the texture to that of the output
+    box->x = 0;
+    box->width = output_width;
+    box->height =
+        (int)(bg_texture->height * output_width / (double)bg_texture->width);
```
Comment (Aki-7) [inline zen/screen-renderer.c:46]:
```suggestion
        bg_texture->height * output_width / bg_texture->width;
```

File: zen/screen-renderer.c [hunk 3]
```diff
@@ -234,9 +275,10 @@ zn_screen_renderer_render(struct zn_screen *screen,
   rects = pixman_region32_rectangles(&screen_damage, &rect_count);
   for (int i = 0; i < rect_count; i++) {
     zn_screen_renderer_scissor_output(output, &rects[i]);
-    wlr_renderer_clear(renderer, (float[]){0.2, 0.3, 0.2, 1});
+    wlr_renderer_clear(renderer, (float[]){1.0, 1.0, 10, 1});
+    zn_screen_renderer_render_background(
+        screen, renderer, server->scene->bg_texture);
```
Comment (Aki-7) [inline zen/screen-renderer.c:280]: How about iterating the damage rectangles in `zn_screen_renderer_render_background`?
All other `zn_screen_renderer_render_*` functions do so.

```c
zn_screen_renderer_render_background(
    struct zn_screen *screen, struct wlr_renderer *renderer, 
    struct wlr_texture *bg_texture, pixman_region32_t *screen_damage) {
  ...
  for (int i = 0; i < rect_count; i ++) {
    zn_screen_renderer_scissor_output(output, &rects[i]);
    wlr_renderer_clear(renderer, (float[]){1.0, 1.0, 10, 1});
    if (bg_texture)
      wlr_render_texture_with_matrix(renderer, bg_texture, matrix, 1.0f);
  }
}
```

**Review by Aki-7 — CHANGES_REQUESTED** [review]
Did all the images display well?
(Sorry I didn't confirm that for myself)

**Comment by liwii** [comment]
Logo
5A11143B-3732-49E7-B5F3-68242BEA56DF
Tricolor
1453BDC0-EB0A-4B41-993C-1E72E5A4D377
Blue
364B4E99-9369-42C6-BB2F-D080491F8C4F
Gray
CB1E919C-7685-4A07-96D9-87BADF9A65E2

They look fine to me

File: zen/scene/scene.c [hunk 4]
```diff
@@ -173,6 +178,31 @@ zn_scene_setup_bindings(struct zn_scene* self)
       zn_scene_new_board_binding_handler, self);
 }
 
+void
```
[the remark this answers is not in the document]
Comment (liwii) [reply zen/scene/scene.c:181]: Thanks, it should be better to mark `static` when possible.

**Comment by liwii** [comment]
Thanks for the review. I fixed some, and I'll deal with the unresolved comments tomorrow.

File: zen/screen-renderer.c (hunk 2 above)
Comment (Aki-7) [reply zen/screen-renderer.c]: Can you also fix the almost identical code a few lines above?

**Comment by Aki-7** [comment]
My suggestions are not meant to be a final code. I would like you to edit them in your editor accordingly. 
There are formatting errors with the current commit. Sorry for the CI misconfiguration (fixed in #158), but I would like you to fix them and rebase to the main branch to apply the new CI configuration.

And I added a new comment https://github.com/zigen-project/zen/pull/152#discussion_r970293056.

File: zen/screen-renderer.c (hunk 2 above)
Comment (liwii) [reply zen/screen-renderer.c]: Done.

File: zen/scene/scene.c [hunk 5]
```diff
@@ -173,6 +178,31 @@ zn_scene_setup_bindings(struct zn_scene* self)
...
+zn_scene_setup_background(struct zn_scene* self, const char* background_png)
+{
+  cairo_surface_t* surface =
+      cairo_image_surface_create_from_png(background_png);
```
[the remark this answers is not in the document]
Comment (liwii) [reply zen/scene/scene.c:184]: It seems that  `wlr_texture_from_pixels` allcates its own buffer and copies the data to it.

https://github.com/swaywm/wlroots/blob/0855cdacb2eeeff35849e2e9c4db0aa996d78d10/render/wlr_texture.c#L32
https://github.com/swaywm/wlroots/blob/9579d62a160821a107763325a515d3aee0a1e158/types/wlr_buffer.c#L483

I would destroy the surface immediately after the wlr texture creation and deal with `wlr_texture` in `zn_scene_destroy`.

File: zen/scene/scene.c [hunk 6]
```diff
@@ -173,6 +178,31 @@ zn_scene_setup_bindings(struct zn_scene* self)
...
+
+  if (surface == NULL) {
```
[the remark this answers is not in the document]
Comment (liwii) [reply zen/scene/scene.c]: Fixed. Thanks for pointing it out.

File: zen/screen-renderer.c (hunk 3 above)
Comment (liwii) [reply zen/screen-renderer.c]: It looks cleaner. Thanks for the suggestion.

**Comment by liwii** [comment]
Now I think the code is ready for review again!!

File: zen/scene/scene.c [hunk 7]
```diff
@@ -174,6 +177,36 @@ zn_scene_setup_bindings(struct zn_scene* self)
...
+  cairo_surface_t* surface =
+      cairo_image_surface_create_from_png(background_png);
+  cairo_t* cr = cairo_create(surface);
+  cairo_status_t status = cairo_status(cr);
+  if (status == CAIRO_STATUS_NO_MEMORY ||
+      status == CAIRO_STATUS_FILE_NOT_FOUND ||
+      status == CAIRO_STATUS_READ_ERROR || status == CAIRO_STATUS_PNG_ERROR) {
+    zn_warn("Background image not loaded");
+    return;
+  }
```
Comment (watasuke102) [inline zen/scene/scene.c:192]: Can you use `if (status != CAIRO_STATUS_SUCCESS)` ?

File: meson.build [hunk 8]
```diff
@@ -85,11 +86,35 @@ wayland_server_dep = dependency('wayland-server', version: wayland_server_req)
...
+    'assets/Zen_Wallpaper_Tricolor_3840x2160.png',
+    'assets/Zen_Wallpaper_Gray_3840x2160.png',
+    'assets/Zen_Wallpaper_Main_3840x2160.png',
+  )
+  wallpaper_install_dir = join_paths(datadir, 'backgrounds', 'zen')
+
+  install_data(wallpaper_files, install_dir: wallpaper_install_dir)
+  default_wallpaper = join_paths('/usr/local', wallpaper_install_dir, 'Zen_Wallpaper_Main_3840x2160.png')
+  cdata.set_quoted('DEFAULT_WALLPAPER', default_wallpaper)
+endif
```
Comment (Aki-7) [inline meson.build:109]: Build fails when `default-wallpaper` option is false.

**Review by Aki-7 — CHANGES_REQUESTED** [review]

File: meson.build [hunk 9]
```diff
@@ -85,11 +86,35 @@ wayland_server_dep = dependency('wayland-server', version: wayland_server_req)
...
+
+config_h = configure_file(
+  input: 'zen-config.h.meson',
+  output: 'zen-config.h',
+  configuration: cdata
+)
+
+zen_inc = include_directories('include', '.')
```
[the remark this answers is not in the document]
Comment (liwii) [reply meson.build]: Thanks, I moved `zen-config.h.meson` to `zen/` and let `zen/meson.build` to directly include the configuration file.

File: meson.build (hunk 8 above)
Comment (liwii) [reply meson.build:111]: Now it assigns "" to `DEFAULT_WALLPAPER` when the option is false (this case should be properly handled in the source code) .

**Review by Aki-7 — APPROVED** [review]
It's motivational!

**Pull request merged**